### PR TITLE
Update description for forwarding-viable

### PR DESCRIPTION
**Pull request by sachendras**
Add more clarity/rules for the forwarding-viable leaf.

1. When a controller "voluntarily" changes the state of all member-links on a LAG bundle or the bundle itself to forwarding-viable=False, it expects that the port/bundle is not used to transmit data-plane traffic even when that is the port of last resort.

a. This requirement can be easily be achieved by an implementation if they follow the rule that, when a port/bundle is set to forwarding-viable=false, do not transmit L3 packets. This is true for both control-plane or data-plane packets. As a result, it would be a graceful transition for L3 traffic on that connection because, the other end of the connection will have its L3 protocol's dead-interval/hold-down timer expired leading for it to migrate away from the connection eventually.

2. I feel minimum-link calculation for aggregate bundle should be ignored for member ports whose forwarding-viable=false. Minimum-Links add unnecessary complications on the controller end. Also, that will result in an implementation to disable a bundle if the min-links requirement is not met and that is not the expectation. Expectation is that, the implementation must continue to Receive and Transmit L2 and only Receive L3 packets on such ports.

3. Finally, in order to remove confusions around IS-IS, the pull calls out that IS-IS should be treated as L3 protocols from this leaf's perspective.

More details in [go/forwarding_viable_expected_behavior](http://goto.google.com/forwarding_viable_expected_behavior)

### CODE REVIEW

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 1]
```diff
@@ -51,7 +51,26 @@ module openconfig-if-sdn-ext {
...
+          1. Given that L2 protocols like LACP should be allowed to be 
+          transmitted and received on such ports, a device mustn't consider 
+          such ports in the calculation for minimum number of links required 
+          for an aggregate bundle when such a configuration exists on the box.
+          Hence, the device musn't disable the bundle because the minimum
+          number of member links requirement is not met for the bundle.
+          2. While it is required that L2 protocols like LLDP, LACP must
+          be allowed for transmit and receive on such ports/bundles, CLNS
+          protocol is an exception and must be treated as per the requirement
+          for L3 protocols below.
```
Comment (LimeHat) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:66]: I'd argue that CLNS _is_ a layer-3 protocol.

But to keep things simple and straightforward, maybe we can just state that IS-IS is considered a L3 protocol for the purposes of this definition?

Comment (sachendras) [reply release/models/interfaces/openconfig-if-sdn-ext.yang]: Done

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 2]
```diff
@@ -51,7 +51,25 @@ module openconfig-if-sdn-ext {
...
+        Forwarding-viable=False:
+          1. Given that L2 protocols like LACP should be allowed to be 
+          transmitted and received on such ports, a device mustn't consider 
+          such ports in the calculation for minimum number of links required 
+          for an aggregate bundle when such a configuration exists on the box.
+          Hence, the device musn't disable the bundle because the minimum
+          number of member links requirement is not met for the bundle.
+          2. L2 protocols like LLDP, LACP must be allowed for transmit and 
+          receive on such ports/bundles, IS-IS should be handled as per the 
+          requirements for L3 protocols below.
```
Comment (dplore) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:65]: This can be removed, it is redundant to text in lines 53-54.

Comment (sachendras) [reply release/models/interfaces/openconfig-if-sdn-ext.yang:62]: I actually updated the section above. Want to keep the rules section clean.

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 3]
```diff
@@ -47,11 +47,26 @@ module openconfig-if-sdn-ext {
...
+        Forwarding-viable=False:
+          1. Given that L2 protocols like LACP should be allowed to be 
+          transmitted and received on such ports, a device mustn't consider 
+          such ports in the calculation for minimum number of links required 
+          for an aggregate bundle when such a configuration exists on the box.
+          Hence, the device mustn't disable the bundle because the minimum
+          number of member links requirement is not met for the bundle.
+          2. L2 protocols like LLDP, LACP must be allowed for transmit and 
+          receive on such ports/bundles, IS-IS should be handled as per the 
+          requirements for L3 protocols below.
```
Comment (rszarecki) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:62]:
```suggestion
          requirements for L3 packets below.
```

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 4]
```diff
@@ -47,11 +47,26 @@ module openconfig-if-sdn-ext {
...
+        Forwarding-viable=False:
+          1. Given that L2 protocols like LACP should be allowed to be 
+          transmitted and received on such ports, a device mustn't consider 
+          such ports in the calculation for minimum number of links required 
+          for an aggregate bundle when such a configuration exists on the box.
+          Hence, the device mustn't disable the bundle because the minimum
+          number of member links requirement is not met for the bundle.
+          2. L2 protocols like LLDP, LACP must be allowed for transmit and 
+          receive on such ports/bundles, IS-IS should be handled as per the 
+          requirements for L3 packets below.
```
Comment (rszarecki) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:62]:
```suggestion
          receive on such ports/bundles, IS-IS PDUs should be handled as per the 
          requirements for L3 packets below.
```

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 5]
```diff
@@ -47,11 +47,26 @@ module openconfig-if-sdn-ext {
...
+          Hence, the device mustn't disable the bundle because the minimum
+          number of member links requirement is not met for the bundle.
+          2. L2 protocols like LLDP, LACP must be allowed for transmit and 
+          receive on such ports/bundles, IS-IS PDUs should be handled as per the 
+          requirements for L3 packets below.
+          3. L3 packets, control-plane as well as data-plane must be received
+          but not transmitted on such ports/bundles. By doing so, expectation
+          is that the other end of the connection ceases to use its respective
+          ports for transmit and receive when either the dead-timer or the
+          hold-down timer of its L3 protocol expires.
```
Comment (dplore) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:67]: We should define what the expectation is for received L3 packets.  I would guess the device is supposed to forward L3 packets, right?    If yes, then stating something like the following might make this more clear?

4. Received L3 packets should be evaluated by the data plane as usual.  For example, when a dataplane packet is received, a lookup should be performed and the packet forwarded as usual.  If an L3 control plane packet is received, it should be processed as usual.

Comment (sachendras) [reply release/models/interfaces/openconfig-if-sdn-ext.yang]: updated as per the internal doc, ptal.

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 6]
```diff
@@ -47,11 +47,34 @@ module openconfig-if-sdn-ext {
...
         (usually part of an aggregate) for the purposes of forwarding
         traffic. This allows a logical aggregate to continue to be
-        used with partial capacity, for example.  Note that setting
-        `forwarding-viable = false` is not equivalent to
-        administratively disabling the interface -- in particular, the
-        interface is expected to participate in L2 protocols such as
-        LLDP or LACP even if it blocked from forwarding traffic.";
+        used with partial capacity. Setting `forwarding-viable = false` is not 
+        equivalent to administratively disabling the interface.
+        Some rules to follow when a port/bundle is set for
```
Comment (dplore) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:52]: Here and throughout I suggest using the OpenConfig terms interface and aggregate : 

```suggestion
        Some rules to follow when an interface or aggregate interface is set for 
```

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 7]
```diff
@@ -47,11 +47,34 @@ module openconfig-if-sdn-ext {
...
+        used with partial capacity. Setting `forwarding-viable = false` is not 
+        equivalent to administratively disabling the interface.
+        Some rules to follow when a port/bundle is set for 
+        Forwarding-viable=False:
+          1. Given that L2 protocols like LACP should be allowed to be 
+          transmitted and received on such ports, a device mustn't consider 
+          such ports in the calculation for minimum number of links required 
+          for an aggregate bundle when such a configuration exists on the box.
+          Hence, the device mustn't disable the bundle because the minimum
+          number of member links requirement is not met for the bundle.
```
Comment (dplore) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:59]: would it be simpler to say 

```
When forwarding-viable is FALSE the interface oper-state should still reflect the L2 status, which when UP and should be treated as such for aggregate interface `min-links` checks.
```

Comment (sachendras) [reply release/models/interfaces/openconfig-if-sdn-ext.yang]: Done

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 8]
```diff
@@ -47,11 +47,31 @@ module openconfig-if-sdn-ext {
...
+          destination post FIB lookup. Also, received control-plane traffic should be 
+          handled normally. Given that the router is not allowed to transmit L3 control-plane 
+          packets on such a port with its FV set to false, if this is a singleton 
+          port, OR if all member interfaces in a bundle are ever set to be FV=False, 
+          expectations are that the remote end of the connection will have its L3 protocol's 
+          dead-interval OR hold-down timer expired taking down the adjacency/peering on 
+          that connection.
+
+          4. An implementation should follow rule#3 even when the subject
+          port/bundle is the last resort of communication for the device.";
```
Comment (dplore) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:74]:
```suggestion
          4. An implementation should follow rule #3 even when the subject
          interface is the last resort of communication for the device.";
```

File: release/models/interfaces/openconfig-if-sdn-ext.yang (hunk 5 above)
Comment (nachikethas) [reply release/models/interfaces/openconfig-if-sdn-ext.yang]: Hi Sach, this statement appears tailored for a protocol like IS-IS. The peer does not _have to_ stop sending traffic to this device on the non-viable ports based on a timeout even if all the trunk members are not-viable. So, can we modify the statement to talk only about DUT and not enforce expectations on the peer? Specifically,
1. Stop L3 transmit for both control and data-plane packets.
2. L3 forwarding should work as usual on Rx without any timeout considerations. The peer can choose to continue sending this packets here indefinitely.

If you prefer, you can call out IS-IS as an example, where the timeout will happen.

Comment (sachendras) [reply release/models/interfaces/openconfig-if-sdn-ext.yang]: @nachikethas I have updated rule#3 to include other use cases, ptal

File: release/models/interfaces/openconfig-if-sdn-ext.yang [hunk 9]
```diff
@@ -47,11 +47,34 @@ module openconfig-if-sdn-ext {
...
+          dead-interval or Hold-down timer of L3 protocols like IS-IS/BGP on the 
+          peer router may expire taking down the adjacency or peering on that 
+          connection. However, the peer may still continue to use the bundle for 
+          forwarding traffic towards the subject router if the peer's forwarding 
+          table is programmed by an external programming entity. In this scenario, 
+          the subject router will continue to recieve traffic indefinitely on 
+          such a bundle.
+
+          4. An implementation should follow rule #3 even when the subject
+          interface is the last resort of communication for the device.";
```
Comment (dplore) [inline release/models/interfaces/openconfig-if-sdn-ext.yang:77]:
```suggestion
        used with partial capacity. Setting `forwarding-viable = false` is not
        equivalent to administratively disabling the interface.
        Some rules to follow when an interface or aggregate interface is set for
        Forwarding-viable=False:
          1. Aggregate interface '/interfaces/interface/aggregation/state/min-links'
           checks should be evaluated based on
          `/interfaces/interface/state/oper-status`.  'min-links' should not be
          affected by the use of forwarding viable.

          2. L2 protocols like LLDP and LACP must be processed normally on
          transmit and receive on such ports/bundles.  IS-IS PDUs should be
          handled as per the requirements for L3 packets below.

          3. L3 packets must not be transmitted on the interface.  
          
          4. Received L3 packets must be processed normally.  Received data-plane
          traffic will continue to forwarded to its destination post FIB lookup.
          Received control-plane traffic must also be processed normally. 
          
          5. It is possible that the dead-interval or hold-down timer of L3
          protocols like IS-IS/BGP on the peer router may expire taking down the
          adjacency or peering on that connection. However, the peer may still
          continue to transmit packets which are received by the local device.  
          These received packet should continue to be processed normally as
          per rule #4 above.

          For example, if the peer's forwarding table is programmed using gRIBI
          by an external controller, the local device will continue to receive
          packets.

          6. An implementation should follow rule #3 even when the subject
          interface on the local device is the last resort of communication for a
          given destination.  For example, the only nexthop for a destination is
          an aggregate interface which has all member interfaces set to
          forwarding-viable = false.  In this scenario all L3 packets for that
          destination will be dropped.";
```

**Review by dplore — APPROVED** [review]
The text makes sense to me as is.  I've suggested some formatting and style changes which I hope makes it easier to read.  PTAL.

**Review by dplore — APPROVED** [review]
LGTM!  @nachikethas can you please give a final review?

**Review by nachikethas — APPROVED** [review]

**Pull request merged**